homeassistant/components/sensor/sma.py

```diff
@@ -0,0 +1,192 @@
+    vol.Optional(CONF_CUSTOM, default={}): vol.Schema({
+        cv.slug: {
+            vol.Required('key'): vol.All(str, vol.Length(min=13, max=13)),
+            vol.Required('unit'): str
+        }})
+}, extra=vol.PREVENT_EXTRA), _check_sensor_schema)
+
+
+def async_setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up SMA WebConnect sensor."""
+    import pysma
+
+    # Combine sensor_defs from the library and custom config
+    sensor_defs = dict(zip(SENSOR_OPTIONS, [
+        (pysma.KEY_CURRENT_CONSUMPTION_W, 'W'),
+        (pysma.KEY_CURRENT_POWER_W, 'W'),
+        (pysma.KEY_TOTAL_CONSUMPTION_KWH, 'kW/h'),
+        (pysma.KEY_TOTAL_YIELD_KWH, 'kW/h')]))
+    for name, prop in config[CONF_CUSTOM].items():
+        if name in sensor_defs:
+            _LOGGER.warning("Custom sensor %s replace built-in sensor", name)
+        sensor_defs[name] = (prop['key'], prop['unit'])
+
+    # Prepare all HASS sensor entities
+    hass_sensors = []
+    used_sensors = []
+    for name, attr in config[CONF_SENSORS].items():
+        hass_sensors.append(SMAsensor(name, attr, sensor_defs))
+        used_sensors.append(name)
+        used_sensors.extend(attr)
+
+    # Remove sensor_defs not in use
+    sensor_defs = {name: val for name, val in sensor_defs.items()
+                   if name in used_sensors}
+
+    yield from add_devices(hass_sensors)
+
+    # Init the SMA interface
+    session = async_get_clientsession(hass)
+    grp = {GROUP_INSTALLER: pysma.GROUP_INSTALLER,
+           GROUP_USER: pysma.GROUP_USER}[config[CONF_GROUP]]
+    sma = pysma.SMA(session, config[CONF_HOST], config[CONF_PASSWORD],
+                    group=grp)
+    print(sma)
+
+    # Ensure we logout on shutdown
+    @asyncio.coroutine
+    def async_close_session(event):
+        """Close the session."""
+        yield from sma.close_session()
+
+    hass.bus.async_listen(EVENT_HOMEASSISTANT_STOP, async_close_session)
+
+    # Read SMA values periodically & update sensors
+    names_to_query = list(sensor_defs.keys())
+    keys_to_query = [sensor_defs[name][0] for name in names_to_query]
+
+    backoff = 0
+
+    @asyncio.coroutine
+    def async_sma(event):
+        """Update all the SMA sensors."""
+        nonlocal backoff
+        if backoff > 1:
+            backoff -= 1
+            return
+
+        print('SMA query:', keys_to_query)
+        values = yield from sma.read(keys_to_query)
+        if values is None:
+            backoff = 3
+            return
+        res = dict(zip(names_to_query, values))
+        print("update sma sensors {} {}".format(values, res))
+        tasks = []
+        for sensor in hass_sensors:
+            task = sensor.async_update_values(res)
+            if task:
+                tasks.append(task)
+        if tasks:
+            yield from asyncio.wait(tasks, loop=hass.loop)
+
+    secs = range(0, 60, config.get(CONF_SCAN_INTERVAL, 5))
+    async_track_utc_time_change(hass, async_sma, second=secs)
```

Comment:
We have a new helpers with timedelta support and now CONF_SCAN_INTERVAL is also a timedelta.